sharktank/tests/models/punet/sharded_resnet_block_with_iree_test.py

```diff
@@ -0,0 +1,247 @@
+# Copyright 2024 Advanced Micro Devices, Inc
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from pathlib import Path
+import tempfile
+
+import torch
+
+
+from shark_turbine import aot
+from sharktank.models.punet.testing import make_resnet_block_2d_theta
+from sharktank.models.punet.layers import ResnetBlock2D
+from sharktank.models.punet.sharding import ResnetBlock2DSplitOutputChannelsSharding
+from sharktank import ops
+from sharktank.types import *
+import iree.runtime
+from typing import List, Optional
+import os
+import logging
+
+logger = logging.getLogger(__name__)
+vm_context: iree.runtime.VmContext = None
+
+
+def get_compiler_args(target_device_kind: str, shard_count: int) -> List[str]:
+    result = [
+        f"--iree-hal-target-device={target_device_kind}[{i}]"
+        for i in range(shard_count)
+    ]
+    return result
+
+
+def compile_iree_module(
+    export_output: aot.ExportOutput, module_path: str, shard_count: int
+):
+    export_output.session.set_flags(
+        *get_compiler_args(target_device_kind="llvm-cpu", shard_count=shard_count)
+    )
+    logger.info(f"Compiling module with flags: {export_output.session.get_flags()}")
+    export_output.compile(save_to=module_path, target_backends=None)
+
+
+def run_iree_module(
+    sharded_input_image: ShardedTensor,
+    sharded_input_time_emb: ShardedTensor,
+    module_path: str,
+    parameters_path: str,
+) -> ShardedTensor:
+    assert sharded_input_image.shard_count == sharded_input_time_emb.shard_count
+    shard_count = sharded_input_image.shard_count
+    logger.info(f"Running module.")
+    hal_driver = iree.runtime.get_driver("local-task")
+    vm_instance = iree.runtime.VmInstance()
+    available_devices = hal_driver.query_available_devices()
+    # Use the same actual device for all devices.
+    devices = [
+        hal_driver.create_device(available_devices[0]) for _ in range(shard_count)
+    ]
```

Comment:
@stellaraccident, you mentioned that we can target multiple IREE CPU devices.
What I have done here is effectively equivalent to
```
iree-run-module \
  --device=local-task \
  --device=local-task \
  ....
```
This sets the assigns `local-task` to the 2 device promises in the model.
Is this what you had in mind?